### PR TITLE
AR-614 Create server for CMS (banners and topics)

### PR DESCRIPTION
**Pull request by gshilin**

### CODE REVIEW

File: api/cms.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"gopkg.in/gin-gonic/gin.v1"
+	"os"
+)
+
+func CMSPerson(c *gin.Context) {
+	id := c.Query("id")
+	fbLang := c.DefaultQuery("fbLang", "en")
+	lang := c.DefaultQuery("lang", fbLang)
+
+	assets := viper.GetString("cms.assets")
```
Comment (edoshor) [inline api/cms.go]: move to common.go as some sort of global

File: api/cms.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"errors"
+	"github.com/spf13/viper"
+	"gopkg.in/gin-gonic/gin.v1"
+	"os"
+)
+
+func CMSPerson(c *gin.Context) {
+	id := c.Query("id")
+	fbLang := c.DefaultQuery("fbLang", "en")
```
Comment (edoshor) [inline api/cms.go]: language and it's fallback are in consts.go

File: api/cms.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	lang := c.DefaultQuery("lang", fbLang)
+
+	assets := viper.GetString("cms.assets")
+
+	file := assets + "persons/persons-" + id + "-" + lang + "-html"
```
Comment (edoshor) [inline api/cms.go]: DRY it up !

File: cmd/cms.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	"net/http"
+	"os"
+	"path"
+	"regexp"
+	"time"
+)
+
+// TODO:
+// 1. Add support for active-passive directories
+
```
Comment (edoshor) [inline cmd/cms.go]: look at events command (cmd package) and its clean separation in events package

File: cmd/cms.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+		}
+		fmt.Println("Persons synced")
+	},
+}
+
+func init() {
+	RootCmd.AddCommand(cmsCmd)
+}
+
+func syncPersons(cms string, assets string) error {
```
Comment (edoshor) [inline cmd/cms.go]: make all this code pretty !!!

File: api/cms.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
+package api
+
+import (
+	"errors"
```
[the remark this answers is not in the document]
Comment (gshilin) [reply api/cms.go]: done

File: .gitignore [hunk 7]
```diff
@@ -10,6 +10,7 @@ sqlboiler.toml
 **/rambler.json
 *.dump.sql
 _vendor-*/
+vendor/**
```
[the remark this answers is not in the document]
Comment (gshilin) [reply .gitignore]: done

File: api/cms.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/spf13/viper"
+	"gopkg.in/gin-gonic/gin.v1"
+	"os"
+)
+
+func CMSPerson(c *gin.Context) {
+	id := c.Query("id")
```
[the remark this answers is not in the document]
Comment (gshilin) [reply api/cms.go]: done

File: api/cms.go (hunk 1 above)
Comment (gshilin) [reply api/cms.go]: done

File: api/cms.go (hunk 2 above)
Comment (gshilin) [reply api/cms.go]: done

File: api/cms.go (hunk 3 above)
Comment (gshilin) [reply api/cms.go]: done

File: cmd/cms.go (hunk 4 above)
Comment (gshilin) [reply cmd/cms.go]: done

File: cmd/cms.go (hunk 5 above)
Comment (gshilin) [reply cmd/cms.go]: done

File: api/routes.go [hunk 9]
```diff
@@ -53,6 +53,12 @@ func SetupRoutes(router *gin.Engine) {
 
 	router.GET("/feeds/morning_lesson", FeedMorningLesson)
 
+	cms := router.Group("/cms")
+	cms.GET("/person", CMSPerson)
```
Comment (edoshor) [inline api/routes.go]: make id a path param ':id'
remove from request struct

File: cms/cms.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+		// create directories for images
+		mkdir(assets+"images/"+path.Dir(image), 0755)
+		saveImage(image, imgURL, assets)
+	}
+}
+
+func mkdir(dirname string, permissions os.FileMode) {
+	if err := os.MkdirAll(dirname, permissions); err != nil {
+		log.Fatalf("Unable to create directory: %s", dirname)
```
Comment (edoshor) [inline cms/cms.go]: log.Fatal is panic. change to fmt.Errorf

File: cms/cms.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+// TODO:
+// 1. Add support for active-passive directories
+
+func SyncCMS() {
+	cms := viper.GetString("cms.url")
+	assets := viper.GetString("cms.assets")
+
+	mkdir(assets+"banners/", 0755)
```
Comment (edoshor) [inline cms/cms.go]: filepath.Join

File: cms/cms.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+	mkdir(assets+"persons/", 0755)
+
+	fmt.Println("Syncing Banners...")
```
Comment (edoshor) [inline cms/cms.go]: use log for all logging

File: cms/cms.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+func mkdir(dirname string, permissions os.FileMode) {
+	if err := os.MkdirAll(dirname, permissions); err != nil {
+		log.Fatalf("Unable to create directory: %s", dirname)
+		utils.Must(err)
+	}
+}
+
+func checkSlug4Language(slug string, pattern string) {
+	regLang, _ := regexp.Compile(pattern)
```
Comment (edoshor) [inline cms/cms.go]: regex that can be compiled as constant should be package level vars.
Use MustCompile

File: cms/cms.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+	x := regLang.FindStringSubmatch(slug)
+	if len(x) == 0 {
+		err := fmt.Errorf("\t- slug MUST include language, but does not")
```
Comment (edoshor) [inline cms/cms.go]: Return errors and let callers decide what to do with them.
Use errors.Wrap

File: cms/cms.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+func saveItem(path string, v interface{}) {
+	m, err := json.Marshal(v)
+	utils.Must(err)
+	err = ioutil.WriteFile(path, m, 0644)
+	utils.Must(err)
+}
+
+func getItem(url string, v interface{}) {
+	var err error
+	client := http.Client{
```
Comment (edoshor) [inline cms/cms.go]: make client a package global.
increase timeout 5

File: cms/cms.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		Timeout: time.Second * 2, // Maximum of 2 secs
+	}
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	utils.Must(err)
+	res, err := client.Do(req)
+	utils.Must(err)
+	body, err := ioutil.ReadAll(res.Body)
```
Comment (edoshor) [inline cms/cms.go]: you must close res.Body (maybe ioutil.ReadAll does that)
defer res.Body.Close

File: api/routes.go (hunk 9 above)
Comment (gshilin) [reply api/routes.go]: done

**Pull request merged**